crates/volta-core/src/tool/mod.rs

```diff
@@ -1,660 +1,188 @@
 }
 
-/// Distinguish global `add` commands in npm or yarn from all others.
-enum CommandArg {
-    /// The command is a *global* add command.
-    GlobalAdd(Option<OsString>),
-    /// The command is a local, i.e. non-global, add command.
-    NotGlobalAdd,
+#[inline]
+fn info_pinned<T: Display + Sized>(tool: T) {
+    info!("{} pinned {} in package.json", success_prefix(), tool);
+}
+
+/// Trait representing all of the actions that can be taken with a tool
+pub trait Tool: Display {
+    /// Fetch a Tool into the local inventory
+    fn fetch(self, session: &mut Session) -> Fallible<()>;
+    /// Install a tool, making it the default so it is available everywhere on the user's machine
+    fn install(self, session: &mut Session) -> Fallible<()>;
+    /// Pin a tool in the local project so that it is usable within the project
+    fn pin(self, session: &mut Session) -> Fallible<()>;
 }
 
 /// Specification for a tool and its associated version.
-///
-/// Since [`Ord`] is implemented for `ToolSpec`, we can use `.sort` on any
-/// `Vec<ToolSpec>`, and the order of the enum variants in the declaration
-/// determines the sorting order, which lets us guarantee (for example) that
-/// Node will always be prioritized over other tools in the toolchain when
-/// dealing with multiple tools.
-///
-/// [`Ord`]: https://doc.rust-lang.org/1.34.0/core/cmp/trait.Ord.html
-#[derive(PartialEq)]
-pub enum ToolSpec {
+#[derive(Debug, PartialEq)]
+pub enum Spec {
     Node(VersionSpec),
     Npm(VersionSpec),
     Yarn(VersionSpec),
     Package(String, VersionSpec),
 }
 
-impl ToolSpec {
-    pub fn from_str_and_version(tool_name: &str, version: VersionSpec) -> Self {
-        match tool_name {
-            "node" => ToolSpec::Node(version),
-            "npm" => ToolSpec::Npm(version),
-            "yarn" => ToolSpec::Yarn(version),
-            package => ToolSpec::Package(package.to_string(), version),
-        }
-    }
+/// A fully resolved Tool, with all information necessary for fetching
+#[derive(Debug)]
+pub enum Resolved {
+    Node(Node),
+    Npm(Npm),
+    Yarn(Yarn),
+    Package(Package),
+}
 
-    pub fn install(&self, session: &mut Session) -> Fallible<()> {
+impl Spec {
+    /// Resolve a tool spec into a fully realized Tool that can be fetched
+    pub fn resolve(self, session: &mut Session) -> Fallible<Resolved> {
         match self {
-            ToolSpec::Node(version) => session.install_node(&version)?,
-            // ISSUE(#292): Implement install for npm
-            ToolSpec::Npm(_version) => throw!(ErrorDetails::Unimplemented {
-                feature: "Installing npm".into()
-            }),
-            ToolSpec::Yarn(version) => session.install_yarn(&version)?,
-            ToolSpec::Package(name, version) => {
-                session.install_package(name.to_string(), &version)?;
+            Spec::Node(version) => {
+                node::resolve(version, session).map(|version| Resolved::Node(Node::new(version)))
```

Comment:
This is totally a take-it-or-leave-it suggestion; I like point-free style but a lot of people don't.

```rust
node::resolve(version, session)
    .map(Node::new)
    .map(Resolved::Node)
```

This works because enum variant constructors are in fact functions!

It ends up reading very nicely and cleanly if you apply this to the following functions as well:

<img width="766" alt="Screen Shot 2019-08-06 at 12 52 37" src="https://user-images.githubusercontent.com/2403023/62567751-14a51f80-b849-11e9-8e83-52cd343268c2.png">


As ever with taste ones, no worries if you prefer it as is, just figured I'd note it as an option.

Reply:
Oh, I like that a lot :+1: